Source/utils/CoordinateTransformations.hpp

```diff
@@ -73,196 +73,226 @@ inverse_spherical_jacobian(const data_t x, const double y, const double z)
-                         double y, double z)
+template <class data_t, class data2_t>
+Tensor<1, data_t, 3>
+cartesian_to_spherical_U(const Tensor<1, data_t, 3> &cartesian_v_U, data_t x,
+                         data2_t y, data2_t z)
 {
-    Tensor<1, data_t> spherical_v_U;
-
-    // derivatives for jacobian matrix - drdx etc
-    Tensor<2, data_t> jac = spherical_jacobian(x, y, z);
-
-    // transform the vector to cartesian coords
-    FOR(i)
-    {
-        spherical_v_U[i] = 0.0;
-        FOR(j) { spherical_v_U[i] += jac[i][j] * cartesian_v_U[j]; }
-    }
-    return spherical_v_U;
+    return transform_vector(cartesian_v_U, spherical_jacobian(x, y, z));
 }
 
 // Convert a vector (with one lower index) in cartesian coords to spherical
 // coords
-template <class data_t>
-Tensor<1, data_t>
-cartesian_to_spherical_L(const Tensor<1, data_t> &cartesian_v_L, data_t x,
-                         double y, double z)
+template <class data_t, class data2_t>
+Tensor<1, data_t, 3>
+cartesian_to_spherical_L(const Tensor<1, data_t, 3> &cartesian_v_L, data_t x,
+                         data2_t y, data2_t z)
 {
-    Tensor<1, data_t> spherical_v_L;
-
-    // derivatives for inverse jacobian matrix - drdx etc
-    Tensor<2, data_t> inv_jac = inverse_spherical_jacobian(x, y, z);
-
-    // transform the vector to cartesian coords
-    FOR(i)
-    {
-        spherical_v_L[i] = 0.0;
-        FOR(j) { spherical_v_L[i] += cartesian_v_L[j] * inv_jac[j][i]; }
-    }
-    return spherical_v_L;
+    return transform_covector(cartesian_v_L,
+                              inverse_spherical_jacobian(x, y, z));
 }
 
 // The area element of a sphere
 template <class data_t>
-data_t area_element_sphere(const Tensor<2, data_t> &spherical_g)
+data_t area_element_sphere(const Tensor<2, data_t, 3> &spherical_g)
 {
     return sqrt(spherical_g[1][1] * spherical_g[2][2] -
                 spherical_g[1][2] * spherical_g[2][1]);
 }
 
+// cartesian coordinates rotation matrix
+// axis must be normalized to 1
+template <class data_t>
+Tensor<2, data_t, 3> rotation_matrix(const Tensor<1, data_t, 3> &axis,
+                                     data_t cos_angle)
+{
+    data_t sine = sqrt(1. - cos_angle * cos_angle);
+    data_t one_minus_cos = 1. - cos_angle;
+
+    // as in
+    // https://en.wikipedia.org/wiki/Rotation_matrix#Rotation_matrix_from_axis_and_angle
+    Tensor<2, data_t, 3> R;
+    R[0][0] = axis[0] * axis[0] * one_minus_cos + cos_angle;
+    R[0][1] = axis[1] * axis[0] * one_minus_cos + axis[2] * sine;
+    R[0][2] = axis[2] * axis[0] * one_minus_cos - axis[1] * sine;
+    R[1][0] = axis[0] * axis[1] * one_minus_cos - axis[2] * sine;
+    R[1][1] = axis[1] * axis[1] * one_minus_cos + cos_angle;
+    R[1][2] = axis[2] * axis[1] * one_minus_cos + axis[0] * sine;
+    R[2][0] = axis[0] * axis[2] * one_minus_cos + axis[1] * sine;
+    R[2][1] = axis[1] * axis[2] * one_minus_cos - axis[0] * sine;
+    R[2][2] = axis[2] * axis[2] * one_minus_cos + cos_angle;
```

Comment:
It is transposed, indeed, and it is done so for the purpose of the Kerr example. ([KerrBH.impl.hpp](https://github.com/GRChombo/GRChombo/blob/main/Source/InitialConditions/BlackHoles/KerrBH.impl.hpp) is the only place where this matrix appears). Unlike the main branch, in here you can provide the directions for the spin in the parameters, and so you have to introduce this rotation matrix to re-define the coordinates when computing the initial data. I am not sure which type of implementation you would prefer to have?

Reply:
I think this may be because I am getting confused by active vs. passive transformations. I think it would be better if the rotation matrix was as on Wikipedia (i.e. the transpose of what it is now) and then the `KerrBH` intial data class was updated as appropriate with comments saying that we want an _active_ transformation hence using the transpose/inverse. for the coordinates.